### DISCUSSION TITLE
Windows builds are here! 🔥

### DISCUSSION THREAD

**Discussion by simonhamp** [NativePHP/laravel#278] [Announcements]
Following on from #271, I'm thrilled to be able to announce that...

- NativePHP now supports Windows! 🪟
- And, Windows builds are working 👷🏽‍♀️
- Plus _everyone_ gets new versions of PHP! 🔥

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExMG90YmlmOXcyY3lnY3A5ZGVldjZrbzFsM2Z2dGRpbnMwanpzNW51eCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/W0BjdrmGjY2aDLyBuw/giphy-downsized.gif)

## Gimme gimme!

The easiest path to upgrade is to run `composer require nativephp/electron` in your app. This will grab the latest version of all the key dependencies.

If you bump into any problems, [please raise an issue](/nativephp/laravel/issues) - but please: **check to see if the same issue has already been reported!**

## Windows support
With [`nativephp/electron` v0.6.0](https://github.com/NativePHP/electron/releases/tag/0.6.0) (and a minor fix in [`nativephp/electron` v0.6.1](https://github.com/NativePHP/electron/releases/tag/0.6.1)), plus the changes in [`nativephp/php-bin` v0.4.0](https://github.com/NativePHP/php-bin/releases/tag/0.4.0), Windows support has _finally_ arrived.

I know many of you have been patiently waiting for this. Over the months, we've racked up quite a few issues as folks have attempted to find some way to get Windows builds working.

I totally get this enthusiasm, Windows support was always on the cards, but we stated early on that Windows support was going to take a while longer for one main reason: a lack of static builds of PHP for Windows with enough extensions baked in to support booting a Laravel application.

Believe me, this wasn't by choice. If we could have launched NativePHP with Windows support, we would have.

Why did it take so long? This is just something that no one has tackled before, until now. And it's not a particularly easy problem to solve.

I don't know if I'm more excited that we have Windows builds or that I was able to go through over half the open tickets and give people some good news.

It's largely thanks to @crazywhalecc and a flock of contributors who have all been working hard on the awesome [`static-php-cli`](/crazywhalecc/static-php-cli) project. [Their efforts](https://github.com/crazywhalecc/static-php-cli/pulls?q=is%3Apr+label%3Aos%2Fwindows+is%3Aclosed) have really paid off. Please go and show them some support!

With that project making the 'building PHP' part easy, I was able to shift attention to making builds of your Electron apps work on Windows. This required some patching and fixing up and _a lot_ of testing, but builds are working reliably enough for the alpha for me to be happy to try and move onto the next thing.

It did also push me to work through many of the tickets and PRs and I finally feel like I'm getting on top of that now.

Thanks for your patience everyone 😊 

## PHP versions aplenty

The other important piece here is wider support for multiple versions of PHP. This has become increasingly important as more and more of you are moving up from PHP 8.1.

**Importantly, this enables full Laravel 11 support as you can now use PHP 8.2 or 8.3 to develop your applications.**

I want NativePHP to make building distributable, desktop-calibre software using PHP a breeze, but there will always be some complexity here.

You see, the version of PHP you use to _develop_ your application is the version that you should _ship_ with your application. If they're not the same, you will get warnings from Composer and potential bugs in your application.

The good news is that you now have complete choice between PHP 8.1, 8.2 and 8.3 and **you don't need to configure anything to make it work**; whatever _minor_ version of PHP you develop with, NativePHP will bundle that same minor version when building your apps.

So if you're building an app on your computer with PHP 8.3.3, NativePHP will select the PHP 8.3 binaries (currently on 8.3.6) to ship when building your app.

Updates to the NativePHP binaries will also happen more regularly.

### A note on `php-bin`
The [`php-bin`](https://github.com/NativePHP/php-bin) package houses the "official" builds of PHP that NativePHP supports. If you choose to build your own statically-compiled versions of PHP to ship with your apps and need a hand, you will need to pay for support.

**The `php-bin` repo will not be accepting PRs containing PHP binaries.** This is out of precaution for the safety of our ecosystem, as we don't have the time and resources to verify the safety and authenticity of such third-party builds.

As we start to move throught the final stages of the alpha towards the beta and general availability stages of NativePHP, we'll also be doing more work to help you verify the safety of the builds _we_ provide. This will be for your benefit and that of your users.

Long term, I don't think we'll be shipping the PHP binaries via this Composer package, as with at least 12 different binaries now bundled in, this one package is becoming a bit of a beast: even zipped, it'll consume over 80MBs of disk space on _each_ app in development.

Of course, when you build your app for distribution only a single one of these binaries is ever embedded inside. Unzipped, they're approx. 25MBs each.

During these early phases of NativePHP's life, this package will suffice, but I expect we'll need to move on from it at some point.

## Super sponsors
While you're here, I want to take this opportunity to say another massive thanks to [our sponsors](https://github.com/sponsors/simonhamp). You've enabled me to spend this time bringing these advancements to everyone. 🙏

But there's still a long long way to go. [As I said back in January](https://github.com/orgs/NativePHP/discussions/248), I'm determined that we get NativePHP into a production-ready state in 2024. While I'm a little bit behind on my rather ambitious "next few weeks" for some of the things mentioned there, I believe we still have plenty of time.

Can you help bring NativePHP to life a little sooner? Please consider [sponsoring](https://github.com/sponsors/simonhamp) - **it doesn't need to be monthly!**

And if you need consultancy or just want some faster support, my sponsorship tiers are a great way for you to get some hands-on help from one of the maintainers of NativePHP!

---

Thanks for coming along for the ride. Please give the latest version of NativePHP a go and report any issues you find.

Once the dust has settled on this release, I will be firing up Tauri again and should start to have some updates for sponsors to try out very soon.

¡Vamos!

**Comment by mtiluk** [comment]
You're awesome.

**Reply by simonhamp** [reply]
Thanks 😊 

Feels like we're getting there!

**Reply by ceau74** [reply]
Thank you so much Simon, congratulations, **Native PHP is amazing**,

**Comment by simonhamp** [comment]
Couple other little bits I forgot to mention:

- We added `sockets` extension support to each version of PHP. This will allow you to build custom TCP-based IPC handling into your application layer so you can allow your application to receive messages from 'outside' - useful if you want to create a kind of 'API' for your app.
- The docs have had a minor spruce-up and some extra details on [Debugging](https://nativephp.com/docs/getting-started/debugging).

**Comment by arnor01** [comment]
Nice work, keep it up!🚀

**Reply by simonhamp** [reply]
Thanks! Will try

**Comment by dlopez525** [comment]
Great job, congratulations. I have a question about this support for Windows, now it supports deep links for Windows

**Reply by simonhamp** [reply]
It should be. I've not tested it, but the logic for deep linking is there

You just need to set the scheme you want to use in your `.env`:

```dotenv
NATIVEPHP_DEEPLINK_SCHEME=myapp
```

Then you can call `myapp://your/app/route` from a browser and it should open your app straight to the relevant screen.

Note that deep linking doesn't work on debug builds on Mac and Linux. You need to run a prod build.

**Comment by Lexian-droid** [comment]
This is very cool, I wanted to give this a try but I was only on windows, happy to finally be able to give it a try! what development stage is it in currently? is it still in alpha like it was in the past?

**Reply by simonhamp** [reply]
Yep just keeping the `alpha` flag until the dust has settled on the Windows release. I just want to do a bit more testing and to monitor the issues for a little while.

Once it feels stable, I think we can move to `beta`.

**Comment by koossaayy** [comment]
Thank you so much ♥

**Reply by simonhamp** [reply]
I got you bro

**Comment by ekramulhasan** [comment]
great innovation and never give up...

**Comment by aliqasemzadeh** [comment]
I am waiting for android and ios.

**Reply by simonhamp** [reply]
As the ARM Mac and Linux builds of PHP are already working here, you will be able to use these on the majority of mobile devices already.

The problem (in theory) is how likely you are to get such apps through App Store review due to the use of an embedded interpreter.

There is [some evidence](https://medium.com/swift2go/embedding-python-interpreter-inside-a-macos-app-and-publish-to-app-store-successfully-309be9fb96a5) to suggest it can pass, but we just haven't had enough examples of that yet.

I'm also not familiar enough with Electron ecosystem tooling to support this.

**Reply by aliqasemzadeh** [reply]
I think we have to make PWA versions and build some API to contact with mobile APIs(Like reading Text Messages and Locations and more)
Do you have build example for Android and iOS?

**Comment by codarck** [comment]
Thanks, I've been waiting for this for months. :c

**Comment by edgaralexanderfr** [comment]
This project looks pretty promising, keep it up @simonhamp !

**Reply by simonhamp** [reply]
Thanks! Gonna do my best

**Comment by mr-ghomi** [comment]
Really great, we really needed something like this.
I hope that laravel reverb can be used soon

**Reply by simonhamp** [reply]
https://github.com/NativePHP/electron-plugin/pull/27

😘

**Comment by bilalswiftsolutions** [comment]
is there anyway  using central database for multiple users so they can work on single database

**Reply by simonhamp** [reply]
This feels like it should be in the [Q&A category](https://github.com/orgs/NativePHP/discussions/categories/q-a)

**Comment by Souvik3798** [comment]
When will we get Andoid and IOS Supports?

**Reply by simonhamp** [reply]
The current builds _already_ support Android (which is just a flavour of Linux) and iOS (which is a variant of the same core as macOS) 🙂

**Comment by Souvik3798** [comment]
Current means alpha or beta?

**Reply by simonhamp** [reply]
Full support for _building_ mobile applications isn't available in NativePHP, but this thread is about Windows builds of the _PHP_ binaries (which are production-ready for all platforms).

Mobile support isn't part of NativePHP yet and likely won't be for v1.

**Comment by Souvik3798** [comment]
Oh..  So we can expect that in maybe the next version?

**Reply by simonhamp** [reply]
I'm guessing this is supposed to be in reply to [this thread](https://github.com/orgs/NativePHP/discussions/278#discussioncomment-10356721)?

Mobile support is likely to take a while. I can't promise any specific delivery timeframe

**Comment by azathub** [comment]
Can I work it on Ubuntu

**Reply by simonhamp** [reply]
Yes, NativePHP should work fine on Ubuntu 🙂

**Comment by hfzm** [comment]
Thank you for this great technology. It really makes it easy to develop a desktop application for Web Developers. I only want to give a feedback on database part. When I seed some data to the database in development and after building and packaging the application the database then does not have the seeded data in built application.

**Reply by koossaayy** [reply]
The database you ship is unique. If you need shared data, you may want to build an API the app can interacts with.

**Reply by simonhamp** [reply]
Or you could seed the user's database on a fresh install

If your seeder also seeded a record that you could check for to see if the seeder was already executed, then it would reliably only run the very first time

**Reply by hfzm** [reply]
Thank you for your reply. I have checked the project folder that is copied with installation in to app directory and checked the database file, seeded data is there but I think it is loading the database from somewhere else and how can we seed on a fresh install. Your help would be appreciated. Thanks a lot.

**Comment by umair982** [comment]
@simonhamp Great work. but its not working correctly on windows. Many people have pointed it. when we hit php artisan native:build win then nothing happens. 
so there are issues for windows. Kindly look in to it.
